feat: create initramfs crate

Roxxas96 opened the pull request:
Implement a crate `initramfs` that pull docker images and wrap them into an initramfs that can be loaded into `lumper`.

sameo commented on initramfs/src/image.rs:
```diff
@@ -0,0 +1,150 @@
+
+        if res.status() != 200 {
+            error!("Failed to get auth token : \n{:?}", res);
+            return Err(anyhow!("Failed to get auth token"));
```
You could add the HTTP result as part of the anyhow error as well.

Roxxas96 replied:
I fixed this while force pushing for the last time

sameo commented on initramfs/src/image.rs:
```diff
@@ -0,0 +1,150 @@
+
+        if res.status() != 200 {
+            error!("Failed to get manifest : \n{:?}", res);
+            return Err(anyhow!("Failed to get manifest"));
```
ditto

Roxxas96 replied:
I fixed this while force pushing for the last time

sameo commented on initramfs/src/image.rs:
```diff
@@ -0,0 +1,150 @@
+            .send()
+            .await?;
+
+        if res.status() != 200 {
```
you only need to check for `res.is_success()`.

Roxxas96 replied:
I fixed this while force pushing for the last time

sameo commented on initramfs/src/image.rs:
```diff
@@ -0,0 +1,150 @@
+        let mut archives = Vec::new();
+        for layer in self.layers.iter() {
+            info!("Pulling layer {:?}", layer.blobSum);
+            let archive = self.pull_layer(layer).await.unwrap();
```
Please avoid `unwrap`ing and propagate a proper error instead.

Roxxas96 replied:
I fixed this while force pushing for the last time

sameo commented on initramfs/src/main.rs:
```diff
@@ -0,0 +1,44 @@
+}
+
+#[tokio::main]
+async fn main() {
```
Your main should return a proper Result, and avoid the following `unwrap`s.

Roxxas96 replied:
I fixed this while force pushing for the last time

RedbeanGit commented on initramfs/src/filesystem.rs:
```diff
@@ -0,0 +1,68 @@
+                }
+
+                // Create a new cpio builder for the file
+                let builder = NewcBuilder::new(entry.header().path().unwrap().to_str().unwrap())
```
You could reuse the `path` variable defined above

Roxxas96 replied:
I fixed this while force pushing for the last time

RedbeanGit commented on initramfs/src/filesystem.rs:
```diff
@@ -0,0 +1,68 @@
+
+                // Create a new cpio builder for the file
+                let builder = NewcBuilder::new(entry.header().path().unwrap().to_str().unwrap())
+                    .uid(entry.header().uid().unwrap() as u32)
```
As you get the header several times, maybe you should create a variable for that

Roxxas96 replied:
I fixed this while force pushing for the last time

RedbeanGit commented on initramfs/src/filesystem.rs:
```diff
@@ -0,0 +1,68 @@
+            // Iterate over all the files in the layer
+            for entry in archive.entries().unwrap() {
+                let mut entry = entry.unwrap();
+                let path = entry.header().path().unwrap().to_str().unwrap().to_string();
```
I think you should prevent using `unwrap` and just return an error

Roxxas96 replied:
I fixed this while force pushing for the last time

RedbeanGit commented on initramfs/src/filesystem.rs:
```diff
@@ -0,0 +1,68 @@
+
+    pub fn write_cpio(&mut self, file_name: &str) {
+        // Write the cpio to disk
+        let archive = File::create(file_name).unwrap();
```
idem, you should prevent using unwrap and handle errors gracefully

Roxxas96 replied:
I fixed this while force pushing for the last time

sameo commented on initramfs/src/image.rs:
```diff
@@ -0,0 +1,172 @@
+        Ok(image)
+    }
+
+    async fn get_auth_token(&mut self) -> Result<()> {
```
nitty nit: I'd call that one `fetch_auth_token`to avoid confusing it with a getter.

Roxxas96 replied:
Ok. I changed it in the last force push !

sameo reviewed (commented):
Just a tiny nit, this looks pretty good to me.

sameo reviewed (commented):
Gentlemen, this PR looks really good, but your git history is a mess. I don't care that at some point you fixed a typo or that you decided to remove some init file. I want to see the structure of your feature, and so most people reading your project git history will. Incidentally, logically articulating your PRs allow for much easier reviews. If you take your time to split your 500 lines of code into 3-4 logically bound commits, this make me (as a reviewer) feel confident that you actually put some thoughts into this work.

So please rework this PR into e.g. 4 commits. This should be easy because your code is actually well architectured (Which is not obvious when I only read the git history):

1. Your tiny http client request.
2. The `Registry` module
3. The `Image` module
4. `main.rs`

sameo commented on initramfs/src/image.rs:
```diff
@@ -0,0 +1,168 @@
+        Ok(image)
+    }
+
+    pub fn get_name(&self) -> &str {
```
getter in Rust are usually not prefixed with `get`. Please rename to `name`

RedbeanGit replied:
fixed!

sameo commented on initramfs/src/image.rs:
```diff
@@ -0,0 +1,168 @@
+        &self.name
+    }
+
+    pub fn get_tag(&self) -> &str {
```
Ditto. s/get_tag/tag/

RedbeanGit replied:
fixed too!

RedbeanGit committed:
feat: add initramfs base files

+ Add Cargo.toml for initramfs crate
* Update .gitignore to prevent versioning of IDE config files
- Remove auto-generated lib.rs

Signed-off-by: RedbeanGit <dubois.julien.mail@gmail.com>

RedbeanGit committed:
feat: add functions to run http request

+ Add httpclient.rs file

Signed-off-by: RedbeanGit <dubois.julien.mail@gmail.com>

RedbeanGit committed:
feat: add Registry and Image structures

+ Add Image structure to represent a docker image
+ Add Registry structure used to run requests to docker registry

Signed-off-by: RedbeanGit <dubois.julien.mail@gmail.com>

RedbeanGit committed:
feat: add main entrypoint for initramfs

+ Add main.rs bringing a cli for initramfs

Signed-off-by: RedbeanGit <dubois.julien.mail@gmail.com>
